### ISSUE TITLE
decouple more view metadata from code

### ISSUE THREAD

**Issue by natechols** [PacificBiosciences/pbreports#135]
The report spec JSON files are a first step, but we are inserting the view metadata as objects are created, which increases code complexity.  A better approach might be adding the view fields at the end:

```
def to_report_with_view(report, report_spec):
    # Overwrite the view attributes and 
    # check if the "schema" is correct (e.g., attribute "alpha" is an int32)
    return new_report
```

since they can be left as None when the objects are initially created.

**Comment by ben-lerch** [comment]
so i'm thinking i'll add `to_report_with_view` as a method of the MetaReport class in report_spec.py. So the workflow will be like:
```my_meta_rpt = MetaReport.from_json('my.json')
#build report w/o metadata
new_report = my_meta_rpt.to_report_with_view(report)```

**Comment by ben-lerch** [comment]
michael, do you have a use case for this? like when we might want multiple views, and where we would specify them. would it be an argument that is fed to the report module that tells it which view to use?

**Comment by natechols** [comment]
The purpose of this is to make it easier to eventually eliminate much of the view metadata from the report JSON files - there's no other use case, just simplifying our code.

**Comment by ben-lerch** [comment]
i thought the idea was that `to_report_with_view` would ensure the report JSON files are *completely* populated with metadata. is that a temporary measure, and eventually the metadata will be added outside of the report modules?

**Comment by natechols** [comment]
This isn't decided yet, but refactoring now will give us more flexibility in the future.

**Comment by mpkocher** [comment]
I believe we want to evolve to, or have a "pure" data report JSON spec where the view and the data isn't coupled. 

However, within the current constraints of the current leaky model, I would suggest this:

```python
a = Attribute("alpha", 1234)
b = Attribute("beta", 456)
r = Report("example", attributes=[a, b])
rx = apply_spec(r, report_spec_view)
return rx
```

This keeps with the current model and adds some useful improvements

- No more view related chatter in the Report. 
- when the `apply_spec` func is called, it should overwrite/set all display related attributes as well as validate the attributes and all report elements (e.g., PlotGroup) exist in the Report.
- We can reuse the spec/view model in the pipeline view rules

This would allow overwriting of values on a per pipeline id basis. 

Also, the minimal view related data in the current report is useful. It doesn't force everyone (internal tools) to have this extra layer, or even implement the View/Spec abstraction to generate a useable report in SL. 

My suggestion is to get the view/spec validation and overwrite the display info in the report and return a new report. Then gut the current python report code for any mention of the display info to remove the unnecessary chatter. 

@natechols Does this seem reasonable?

For a pure data only report, we can meet to discuss this. Mark E is already suggested something similar for the RunQC reports.

**Comment by ben-lerch** [comment]
```validate the attributes and all report elements (e.g., PlotGroup) exist in the Report```

what i'm writing now is a method of metareport that just takes whatever report it gets fed and adds metadata to it. it will only break if the report has an id that doesn't match anything in the metareport. do we want a function that actually enforces the entire view? so, for example, if there's an attribute in the view that isn't in the report then it will say: "error: missing attribute"?

**Comment by natechols** [comment]
> do we want a function that actually enforces the entire view

Yes, it should definitely raise an exception if the report contains anything that isn't part of the spec.  (But not necessarily the reverse, since there may be cases where we deliberately leave out some parts of the report depending on input data.)

**Comment by ben-lerch** [comment]
One complication to this is situations where the id is variable, as in the variants report module:
https://github.com/PacificBiosciences/pbreports/blob/master/pbreports/report/variants.py#L192
How do we want to deal with this?

**Comment by ben-lerch** [comment]
another thought: perhaps `apply_view` should actually enforce type rather than check type. so if a number is supposed to be an int it just gets wrapped in `int()` when the view is applied. i frequently get errors like this: 
`AssertionError: 0.9465 != float`
and I'm not really sure what to make of them but I can avoid them by enforcing type through `apply_view`.

**Comment by natechols** [comment]
You need to change that AssertionError to show the actual type instead of just the value - it's impossible to tell what's breaking otherwise.

**Comment by ben-lerch** [comment]
Good call. So in this case: `AssertionError: <type 'numpy.float64'> != float`

**Comment by ben-lerch** [comment]
this is the latest version. i'll take any feedback:
https://github.com/ben-lerch/pbreports/blob/metadata/pbreports/report/report_spec.py#L181
